src/main/java/build/buildfarm/instance/shard/RedisShardSubscription.java

```diff
@@ -52,11 +71,66 @@ class RedisShardSubscription implements Runnable {
     this.client = client;
   }
 
+  class SubscriptionAutoCloseable implements AutoCloseable {
+    public void initiateSubscription() {
+      SettableFuture subscribeFuture = SettableFuture.create();
+      subscriber.setSubscribeFuture(subscribeFuture);
+      manageState(SubscriptionAction.START_SUBSCRIBE);
+    }
+
+    @Override
+    public void close() {
+      subscriber.setSubscribeFuture(null);
+      manageState(SubscriptionAction.END_SUBSCRIPTION);
+    }
+  }
+
+  private synchronized void manageState(SubscriptionAction update) {
+    SubscriptionState currentState = subscriptionState.get();
+    switch (update) {
+      case STOP:
+        if (currentState == SubscriptionState.NOT_SUBSCRIBED) {
+          subscriptionState.set(SubscriptionState.FULLY_STOPPED);
+        } else if (currentState == SubscriptionState.SUBSCRIBING
+            || currentState == SubscriptionState.SUBSCRIBED) {
+          subscriptionState.set(SubscriptionState.STOPPED_BUT_SUBSCRIBED);
+        }
+        break;
+      case START_SUBSCRIBE:
+        if (currentState != SubscriptionState.STOPPED_BUT_SUBSCRIBED
+            && currentState != SubscriptionState.FULLY_STOPPED) {
+          subscriptionState.set(SubscriptionState.SUBSCRIBING);
+        }
+        break;
+      case END_SUBSCRIPTION:
+        if (currentState == SubscriptionState.STOPPED_BUT_SUBSCRIBED
+            || currentState == SubscriptionState.FULLY_STOPPED) {
+          subscriptionState.set(SubscriptionState.FULLY_STOPPED);
+        } else {
+          subscriptionState.set(SubscriptionState.NOT_SUBSCRIBED);
+        }
+        break;
+    }
+  }
+
   private void subscribe(UnifiedJedis jedis, boolean isReset) {
     if (isReset) {
       onReset.accept(jedis);
     }
-    jedis.subscribe(subscriber, subscriptions.get().toArray(new String[0]));
+
+    try (SubscriptionAutoCloseable subscriptionAutoCloseable = new SubscriptionAutoCloseable()) {
+
+      subscriptionAutoCloseable.initiateSubscription();
```

Comment:
This behavior should be a part of the creation of the auto closable, not a separate method.
Make it a part of the constructor, a static on the class, or a method on the Subscription class to produce it (and initiate the transaction).

Reply:
I did this for the admittedly hacky reason that if all the action takes place in the constructor then the linter will complain that `subscriptionAutoCloseable` is an unused variable.
In my last commit https://github.com/buildfarm/buildfarm/pull/1877/commits/e4f05f63009765583634343ce8a5b3be7e14b959 I only used the closeable for handling the `close()` and I got that complaint from the linter. Is there a way to add an ignore/suppress, or can you suggest a better way to add a "use" for the variable?

Reply:
You can just instantiate the object with no variable name, it will still auto-close, and this will likely defeat the linter.